src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -517,6 +517,62 @@ private static void ArrayTypeCheck_Helper(object obj, void* elementType)
             }
         }
 
+        // Helpers for boxing
+        [DebuggerHidden]
+        private static object? Box_Nullable(IntPtr handle, ref byte nullableData)
+        {
+            MethodTable* srcMT = new TypeHandle((void*)handle).AsMethodTable();
+            return BoxNullableImpl(srcMT, ref nullableData);
+        }
+
+        [DebuggerHidden]
+        internal static object? BoxNullableImpl(MethodTable* srcMT, ref byte nullableData)
+        {
+            Debug.Assert(srcMT->IsNullable);
+
+            // If 'hasValue' is false, return null.
+            if (!Unsafe.As<byte, bool>(ref nullableData))
+                return null;
+
+            // Allocate a new instance of the T in Nullable<T>.
+            MethodTable* dstMT = srcMT->InstantiationArg0();
+            ref byte srcValue = ref Unsafe.Add(ref nullableData, srcMT->NullableValueAddrOffset);
+
+            // Delegate to non-nullable boxing implementation
+            return BoxImpl(dstMT, ref srcValue);
+        }
+
+        [DebuggerHidden]
+        internal static object Box(IntPtr handle, ref byte unboxedData)
+        {
+            MethodTable* typeMT = new TypeHandle((void*)handle).AsMethodTable();
+            Debug.Assert(typeMT->IsValueType && !typeMT->IsNullable);
+
+            return BoxImpl(typeMT, ref unboxedData);
+        }
+
+        [DebuggerHidden]
+        internal static object BoxImpl(MethodTable* typeMT, ref byte unboxedData)
+        {
+            Debug.Assert(typeMT != null);
+            Debug.Assert(typeMT->IsValueType);
+
+            object boxed = RuntimeTypeHandle.GetThreadLocalAllocInternal(typeMT) ??
+                RuntimeTypeHandle.InternalAllocNoChecks(typeMT);
+
+            if (typeMT->ContainsGCPointers)
+            {
+                Buffer.BulkMoveWithWriteBarrier(ref boxed.GetRawData(), ref unboxedData,
+                    typeMT->GetNumInstanceFieldBytesIfContainsGCPointers());
+            }
+            else
+            {
+                SpanHelpers.Memmove(ref boxed.GetRawData(), ref unboxedData, typeMT->GetNumInstanceFieldBytes());
+            }
```

Comment:
We can unroll it manually if this is the bottleneck
 ```suggestion
            else
            {
                int bytes = (int)typeMT->GetNumInstanceFieldBytes();
                switch (bytes)
                {
                    case 1:
                        Unsafe.As<byte, byte>(ref dst) = Unsafe.As<byte, byte>(ref src);
                        break;
                    case 2:
                        Unsafe.As<byte, ushort>(ref dst) = Unsafe.As<byte, ushort>(ref src);
                        break;
                    case 4:
                        Unsafe.As<byte, uint>(ref dst) = Unsafe.As<byte, uint>(ref src);
                        break;
                    case 8:
                        Unsafe.As<byte, ulong>(ref dst) = Unsafe.As<byte, ulong>(ref src);
                        break;
                    default:
                        SpanHelpers.Memmove(ref dst, ref src, (nuint)bytes);
                        break;
                }
            }
```

Reply:
Applied it for testing: https://github.com/EgorBot/runtime-utils/issues/348#issuecomment-2839896783

Reply:
@filipnavara, it showed a slight improvement, but still this is not it. 🥲

Reply:
Is `InternalAllocNoChecks` getting inlined? You can try to mark it no-inline

Reply:
It shows some interesting data 👍 I was mainly wondering whether the overhead is in the allocation or the copying. The code paths for small sizes improved to the point that 2/3 of the platforms are nearly on par so that seems to point to the copying being the primary bottleneck.